src/tunnel.cc

```diff
@@ -898,6 +898,17 @@ void
 TunnelStateData::copyRead(Connection &from, IOCB *completion)
 {
     assert(from.len == 0);
```

Comment:
I checked that this assertion is valid: copyRead() is called via writeServerDone() -> copyClientBytes(), and  client.dataSent() in writeServerDone() resets this field. Also for the initial case (when copyRead() is not called via writeServerDone()), copyClientBytes() takes care about the 'pre read data' in a different execution path, and from.len is not checked in this case. So I think the code below can simply close the 'other' connection, we don't need finishWritingAndDelete() with its "waiting to finish writing" logic there.

Reply:
> I checked that this assertion is valid: copyRead() is called via ...

AFAICT, you have checked that this assertion is not going to fail in current code. I see several red flags in the description of those checks, but let's assume that it does not fail indeed. That is great, but it does not help me understand how new code (placed below this assertion) relies on the asserted condition...

> ... [complex reasoning about from.len value and possibly other conditions] ... So I think the code below can simply close the 'other' connection, we don't need finishWritingAndDelete() with its "waiting to finish writing" logic there.

I am having trouble connecting the dots here. Would the following statement be an accurate representation of proposed/new code logic? 

* Zero from.len implies that we are not currently writing to toConn. Thus, finishWritingAndDelete(toConn) is unnecessary and `toConn->close()` is sufficient to correctly end transaction and destroy TunnelStateData.


N.B. Official comm_read() calling code below the assertion relies on the asserted condition because it uses a pointer to the beginning of `from.buf` buffer. The number of currently buffered bytes must be zero for comm_read() to safely fill that area of the buffer. From that code point of view, the assertion is correct (i.e. it checks the correct invariant that the code below the assertion relies on) _even if it fails_ (due to Squid bugs elsewhere)!

Reply:
> Zero from.len implies that we are not currently writing to toConn.

We may imply that: from.len becomes non-zero when we have read some new data and are going to write it (Connection::bytesIn() in TunnelStateData::readClient()). It becomes zero again when we have written all available data (Connection::dataSent() asserts on that) and are going to read again (our TunnelStateData::copyRead()). So, len==0 here means that we have written all available data (if any) and do not have new data to write (yet). So when we observe that client has gone (no new data) at this point, we simply close the opposite connection.